### ISSUE TITLE
[Bug] test_multithreading in test_advanced.py hangs when run on windows

### ISSUE THREAD

**Issue by mattip** [ray-project/ray#20173]
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### What happened + What you expected to happen

Running `test_multithreading` in `python/ray/tests/test_advanced' consistently fails on windows. The test runs twice: once in worker mode and once in driver mode. In worker mode, all the subtests fail. In driver mode, only the `test_wait` subtest fails.

The test has no timeout, so it hangs. On CI there is a timeout-killer, but would it be worthwhile to add timeouts to the test itself?

### Versions / Dependencies

windows 10, python 3.8

### Reproduction script

`python -m pytest python\ray\tests\test_advanced.py -k test_multithreading`

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

**Comment by scv119** [comment]
:( @mattip where you able to get a stacktrace when it hangs?

**Comment by mattip** [comment]
@scv119 after updating to e5e62d89918835df5b6823a08c70a45caec5c8c6 it no longer hangs, just fails when using Actors in workers. There are 6 `worker*.err` files in the logs, three of them are empty and three end with
```
:task_name:echo
:task_name:echo
:task_name:echo
:actor_name:Echo
Windows fatal exception: access violation
```

There are corresponding `worker*.out` files that are also empty or end with
```
:task_name:echo
:task_name:echo
:task_name:echo
:task_name:echo
:task_name:run_tests_in_worker
:actor_name:MultithreadedActor
```

Is there code repressing the dialog box usually associated with application crashes for windows? It is difficult to attatch to the correct worker process.

**Comment by mattip** [comment]
It seems the crash is when touching `logging_provider_` in the destructor for the worker. If I comment out the debug print in the worker destructor, I still get a similar crash in `RayLog::~RayLog()` when it tries to delete `logging_provider_`.

**Comment by mattip** [comment]
I think the crash is a false flag. It happens long after the remote objects fail to execute. [Here is a test sequence](https://gist.github.com/mattip/b206b3925fa6a87cddb10ec68ac16d52) that demonstrates the process. It creates a number of threads and calls actors in them. Then it tries to create remote objects and wait for their execution. The remote objects are __never created__ so the test hangs. Here are the log files, after applying #20426. 

[logs.zip](https://github.com/ray-project/ray/files/7547645/logs.zip)
 
The print statements show the problem, after `creating object` there should be a print of the remote objects running the `echo` function, but none ever appears:
```
:task_name:run_tests_in_worker
---run_test_in_multi_threads test_call_actor num_threads=10 num_repeats=25 ---
creating objects
---run_test_in_multi_threads test_wait num_threads=10 num_repeats=1 ---
```

**Comment by wuisawesome** [comment]
@scv119 do you mind helping out and looking into this? i'm wondering if this is related to the other multi-threaded actor issues.

**Comment by mattip** [comment]
For convenience, here is the guts of the test. It calls a function `test_api_in_multi_threads` from a worker, which in turn calls `run_test_in_multi_threads(func)` with two functions: one to call actions via actors, and then another to wait for remote calls.

```
    def run_test_in_multi_threads(test_case, num_threads=10, num_repeats=25):
        """A helper function that runs test cases in multiple threads."""

        name = str(test_case).rsplit('.', 1)[-1].split(' ', 1)[0]
        print(f'---run_test_in_multi_threads {name} {num_threads=} {num_repeats=} ---')
        def wrapper():
            for _ in range(num_repeats):
                test_case()
                time.sleep(random.randint(0, 10) / 1000.0)
            return "ok"

        executor = ThreadPoolExecutor(max_workers=num_threads)
        futures = [executor.submit(wrapper) for _ in range(num_threads)]
        for future in futures:
            assert future.result() == "ok"

    @ray.remote
    def echo(value, delay_ms=0):
        if delay_ms > 0:
            time.sleep(delay_ms / 1000.0)
        if value < 100:
            print(f'echo({value},{delay_ms})')
        return value

    def test_api_in_multi_threads():
        """Test using Ray api in multiple threads."""

        @ray.remote
        class Echo:
            def echo(self, value):
                return value

        # Test multiple threads calling one actor.
        actor = Echo.remote()

        def test_call_actor():
            value = random.randint(0, 1000000)
            result = ray.get(actor.echo.remote(value))
            assert value == result

        run_test_in_multi_threads(test_call_actor)

        print('creating objects')
        # Test multiple threads waiting for objects.
        num_wait_objects = 10
        objects = [
            echo.remote(i, delay_ms=10) for i in range(num_wait_objects)
        ]

        def test_wait():
            ready, _ = ray.wait(
                objects,
                num_returns=len(objects),
                timeout=1000.0,
            )
            assert len(ready) == num_wait_objects
            assert ray.get(ready) == list(range(num_wait_objects))

        run_test_in_multi_threads(test_wait, num_repeats=1)

    # Run tests in a worker.
    @ray.remote
    def run_tests_in_worker():
        test_api_in_multi_threads()
        return "ok"

    assert ray.get(run_tests_in_worker.remote()) == "ok"

```

**Comment by scv119** [comment]
@mattip yes let me take a look. btw do you have a repro on linux as well?

**Comment by mattip** [comment]
I have not found a way to trigger this failure on linux. Reducing the thread count and repeat to a minimum does not help the situation on windows, and making them larger does not cause a failure on linux (as far as I can tell)

**Comment by mattip** [comment]
The `test_actor_call_order` test in `test_basic_2.py` also exhibits this pattern. If I 
- add a print in the Actor function `inc`
- add a sleep(large number) to the end of the function to not clutter the logs with shutdown messages
- filter out the multiple messages about metric reporting (this PR would help with reducing the noise #20426)
- reduce the num_cpus to 2 `num_cpus=2`

On windows I get ~9 prints from the `inc` function, on linux I get all 100. On both, 4 workers are started (there are both `python-core-worker-NNN-XXX.log` files and `worker-NNN-010000000-XXX.out` files, where `NNN` is the worker id and `XXX` is the PID for each of the 4 workers). One is the actor and echos the prints in the `inc` function. On windows, only one worker gets the remote `small_value` task, on linux they are split between two workers.

The strange thing is that on windows, the python-core-worker with the same PID as the single worker that is performing the `small_value` task gets these lines almost immediately after starting:
```
core_worker.cc:840: Exit signal received, this process will exit after all outstanding tasks have finished, exit_type=IDLE_EXIT
raylet_client.cc:150: RayletClient::Disconnect, exit_type=IDLE_EXIT, has creation_task_exception_pb_bytes=0
core_worker.cc:365: Removed worker XXX
```

The other two workers' `*.out` are empty, and the `python-core-worker-*.log` shows only the worker initialization sequence.

On linux, two of the workers (there are two splitting the `small_value` tasks between them) gets this message, one of them is indeed idle and the other stops processing `small_value` tasks. Maybe the message is sent to the wrong workers, on linux this doesn't matter since there are two workers sharing the task but on windows the removal of the active worker prevents the tasks from completing?

**Comment by mattip** [comment]
Maybe the `is_idle` shutdown is a symptom of the problem, not the cause. If I disable this by forcing `is_idle = false;` I still get only 8-9 tasks completed on windows.

**Comment by mattip** [comment]
> The test_actor_call_order test in test_basic_2.py also exhibits this pattern. 

If I change the test and turn the normal remote task into an actor remote task the test passes. So something about mixing normal tasks and actor-based tasks?

**Comment by mattip** [comment]
The problem is two-fold:
- The worker is leased to the actor, but the timeout expires long before it can complete the tasks on windows. To execute 100 tasks on a single CPU requires 6000ms on my machine, and the default is 500ms: https://github.com/ray-project/ray/blob/b872fdaaac997db0b06ec6bc9de0148d714f18ad/src/ray/common/ray_config_def.h#L158-L160
- There is a fallback path which calls `CoreWorkerDirectTaskSubmitter::StealTasksOrReturnWorker`, but this does not seem to be "stealing" the tasks in a way that makes them run later.

A small work around would be to set a longer lease timeout for this test. A larger work around would be to set a default lease timeout of 6000ms on windows.

**Comment by mattip** [comment]
Closed by #20986
